compiler-access/src/test/scala/org/alephium/ralph/lsp/access/compiler/parser/soft/ast/TestSoftAST.scala

```diff
@@ -80,177 +294,401 @@ object TestSoftAST {
+
   def Extends(index: SourceIndex): SoftAST.TokenDocumented[Token.Extends.type] =
     TokenDocumented(
       index = index,
       token = Token.Extends
     )
 
-  def Pub(index: SourceIndex): SoftAST.TokenDocumented[Token.Pub.type] =
+  def Implements(code: String): SoftAST.TokenDocumented[Token.Implements.type] =
+    Implements(indexOf(code))
+
+  def Implements(index: SourceIndex): SoftAST.TokenDocumented[Token.Implements.type] =
     TokenDocumented(
       index = index,
-      token = Token.Pub
+      token = Token.Implements
     )
 
-  def For(index: SourceIndex): SoftAST.TokenDocumented[Token.For.type] =
+  def Using(code: String): SoftAST.TokenDocumented[Token.Using.type] =
+    Using(indexOf(code))
+
+  def Using(index: SourceIndex): SoftAST.TokenDocumented[Token.Using.type] =
     TokenDocumented(
       index = index,
-      token = Token.For
+      token = Token.Using
     )
 
-  def While(index: SourceIndex): SoftAST.TokenDocumented[Token.While.type] =
+  def Emit(code: String): SoftAST.TokenDocumented[Token.Emit.type] =
+    Emit(indexOf(code))
+
+  def Emit(index: SourceIndex): SoftAST.TokenDocumented[Token.Emit.type] =
     TokenDocumented(
       index = index,
-      token = Token.While
+      token = Token.Emit
     )
 
-  def Dot(index: SourceIndex): SoftAST.TokenDocumented[Token.Dot.type] =
+  def Embeds(code: String): SoftAST.TokenDocumented[Token.Embeds.type] =
+    Embeds(indexOf(code))
+
+  def Embeds(index: SourceIndex): SoftAST.TokenDocumented[Token.Embeds.type] =
     TokenDocumented(
       index = index,
-      token = Token.Dot
+      token = Token.Embeds
     )
 
-  def Return(index: SourceIndex): SoftAST.TokenDocumented[Token.Return.type] =
+  def True(code: String): SoftAST.TokenDocumented[Token.True.type] =
+    True(indexOf(code))
+
+  def True(index: SourceIndex): SoftAST.TokenDocumented[Token.True.type] =
     TokenDocumented(
       index = index,
-      token = Token.Return
+      token = Token.True
     )
 
-  def LessThanOrEqual(index: SourceIndex): SoftAST.TokenDocumented[Token.LessThanOrEqual.type] =
+  def False(code: String): SoftAST.TokenDocumented[Token.False.type] =
+    False(indexOf(code))
+
+  def False(index: SourceIndex): SoftAST.TokenDocumented[Token.False.type] =
     TokenDocumented(
       index = index,
-      token = Token.LessThanOrEqual
+      token = Token.False
+    )
+
+  def AlphLowercase(code: String): SoftAST.TokenDocumented[Token.AlphLowercase.type] =
+    AlphLowercase(indexOf(code))
+
+  def AlphLowercase(index: SourceIndex): SoftAST.TokenDocumented[Token.AlphLowercase.type] =
+    TokenDocumented(
+      index = index,
+      token = Token.AlphLowercase
+    )
+
+  def AlphUppercase(code: String): SoftAST.TokenDocumented[Token.AlphUppercase.type] =
+    AlphUppercase(indexOf(code))
+
+  def AlphUppercase(index: SourceIndex): SoftAST.TokenDocumented[Token.AlphUppercase.type] =
+    TokenDocumented(
+      index = index,
+      token = Token.AlphUppercase
```

Comment:
Did you used AI to generate all this? otherwise congrats ^^

Reply:
Nah. Just `println`

```scala
def code(token: Token) = {
  val objectName = token.productPrefix
  s"""
     |def $objectName(code: String): SoftAST.TokenDocumented[Token.$objectName.type] =
     |  $objectName(indexOf(code))
     |
     |def $objectName(index: SourceIndex): SoftAST.TokenDocumented[Token.$objectName.type] =
     |  TokenDocumented(
     |    index = index,
     |    token = Token.$objectName
     |  )
     |
     |""".stripMargin
}

val strings = tokens map code

println(strings.mkString)

```